lib/elixir/lib/map.ex

```diff
@@ -93,11 +96,10 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.reduce([], &[transform.(&1) | &2])
+    |> :lists.reverse
+    |> :maps.from_list
```

Comment:
Collectable for maps could accumulate a list and call `:maps.from_list/1` instead of accumulating a map if that's faster.